group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,399 @@
+
+        // Sorts partitions in ascending order by number of potential consumers with matching racks.
+        // Partitions with no potential members aren't included in this list.
+        List<TopicIdPartition> sortedPartitions = rackInfo.sortPartitionsByRackConsumers(unassignedPartitions);
+
+        Iterator<TopicIdPartition> partitionIterator = sortedPartitions.iterator();
+        while (partitionIterator.hasNext()) {
+            TopicIdPartition partition = partitionIterator.next();
+            boolean assigned = false;
+            for (int i = 0; i < roundRobinMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                Integer remainingPartitionCount = unfilledMembers.get(memberId);
+
+                if (remainingPartitionCount != null && remainingPartitionCount > 0 && !rackInfo.racksMismatch(memberId, partition)) {
+                    assignPartitionToMember(memberId, partition);
+                    assigned = true;
+                    partitionIterator.remove();
+                    unassignedPartitions.remove(partition);
+                }
+
+                // Only re-add to the end of the queue if it's still in the unfilledMembers map
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+        }
+    }
+
+    /**
+     * Allocates the unassigned partitions to available members.
+     *
+     * If the rack-aware strategy is enabled, partitions are attempted to be assigned back to their current owners first.
+     *
+     * If a partition couldn't be assigned to its current owner due to quotas or
+     * if the rack-aware strategy is not enabled, the partitions are allocated to members in a round-robin fashion.</p>
+     */
+    private void unassignedPartitionsRoundRobinAssignment() {
+        Queue<String> roundRobinMembers = new LinkedList<>(unfilledMembers.keySet());
+        Iterator<TopicIdPartition> partitionIterator = unassignedPartitions.iterator();
+
+        while (partitionIterator.hasNext()) {
+            TopicIdPartition partition = partitionIterator.next();
+            boolean assigned = false;
+
+            if (useRackAwareStrategy && currentPartitionOwners.containsKey(partition)) {
+                String prevOwner = currentPartitionOwners.get(partition);
+                if (unfilledMembers.containsKey(prevOwner)) {
+                    assignPartitionToMember(prevOwner, partition);
+                    assigned = true;
+                    partitionIterator.remove();
+                    if (!unfilledMembers.containsKey(prevOwner)) {
+                        roundRobinMembers.remove(prevOwner);
+                    }
+                }
+            }
+
+            // Only re-add the member to the end of the queue if it's still available for assignment.
+            for (int i = 0; i < unfilledMembers.size() && !assigned; i++) {
```

Comment:
how come we iterate on unfilledMembers instead of roundRobinMembers as in `rackAwareRounRobinAssignment`? this looks like a bug; can we add a test that breaks this

Reply:
we are iterating over the roundRobin members, we're just using the size of all the unfilled members so that we do at least n number of iterations

Reply:
cause we want to iterate through/visit every member at least once, n is the number of members.

Reply:
no, we could have removed members from unfilledMembers in the if block above. and if we do so, we would also remove them from roundrobinMembers.

Since they start with the same members, they will end up with the same members

Reply:
1) It doesn't matter if round robin members is not empty, we're doing this by partition so it's very possible that after iterating through the list many members are yet to receive partitions. The only time round robin members will be empty/there's no more members that need partitions is for the last partition.

Reply:
Unless round robin members is empty how will we get a NPE, we would keep polling and adding it back right even if its just one member

Reply:
anyways I changed it to round robin members size just like the rack aware round robin